modin/core/dataframe/algebra/binary.py

```diff
@@ -97,35 +136,100 @@ def compute_dtypes_common_cast(first, second) -> np.dtype:
             ),
         ]
     )
-    dtypes = dtypes.sort_index()
     return dtypes
 
 
-def compute_dtypes_boolean(first, second) -> np.dtype:
+def maybe_build_dtypes_series(
+    first, second, dtype, trigger_computations=False
+) -> Optional[pandas.Series]:
     """
-    Precompute data types for boolean operations.
+    Build a ``pandas.Series`` describing dtypes of the result of a binary operation.
 
     Parameters
     ----------
     first : PandasQueryCompiler
         First operand for which the binary operation would be performed later.
-    second : PandasQueryCompiler
+    second : PandasQueryCompiler, list-like or scalar
         Second operand for which the binary operation would be performed later.
+    dtype : np.dtype
+        Dtype of the result.
+    trigger_computations : bool, default: False
+        Whether to trigger computation of the lazy metadata for `first` and `second`.
+        If False is specified this method will return None if any of the operands doesn't
+        have materialized columns.
 
     Returns
     -------
-    dtypes
-        The pandas series with precomputed dtypes.
+    pandas.Series or None
+        The pandas series with precomputed dtypes or None if there's not enough metadata to compute it.
 
     Notes
     -----
     Finds a union of columns and finds dtypes for all these columns.
     """
+    if not trigger_computations:
+        if not first._modin_frame.has_columns_cache:
+            return None
+
+        if (
+            isinstance(second, type(first))
+            and not second._modin_frame.has_columns_cache
+        ):
+            return None
+
     columns_first = set(first.columns)
-    columns_second = set(second.columns)
-    columns_union = columns_first.union(columns_second)
-    dtypes = pandas.Series([np.dtype(bool)] * len(columns_union), index=columns_union)
-    dtypes = dtypes.sort_index()
+    if isinstance(second, type(first)):
+        columns_second = set(second.columns)
+        columns_union = columns_first.union(columns_second)
+    else:
+        columns_union = columns_first
```

Comment:
we take keys intersection at the front-end
https://github.com/modin-project/modin/blob/baa860554febd58b7aa11947028c367664488887/modin/pandas/base.py#L309-L315

Reply:
I see inconsistency in the handling of different types. I wonder if it can be fixed.
Why not take the intersection of the keys in the case when `other` is a dataframe or a series?

Reply:
Well, I blindly inherited this logic from #5494, and now it seems that it's incorrect. There's indeed a difference in how boolean and non-boolean binary operations treat columns mismatching.

At the first glance, there are 3 types of behavior in pandas:
1. Union columns
2. Raise an error if columns are not fully matching
3. Apply a binary operation to an intersection of columns and fill the rest with NaNs

Here's how each bin op behaves in pandas:
```
combine_first: union of columns in the result

eq : failure if columns don't fully match
ge
gt
le
lt
ne

__and__ : the result is an intersection of columns + nan columns for the rest
__or__
__xor__
... other non-boolean methods
```

So, apparently, not all boolean operations go through the 'union' way, I'll check how to align modin's logic with pandas

Reply:
oh, it seems to be even more complicated, in modin both public and private boolean methods (for example `.eq` and `.__eq__`) go through the same logic, however in pandas they're performing differently:
```python
>>> df1.eq(df2)
       a      b      c
0  False  False  False
1  False  False  False
>>> df1.__eq__(df2) # same as df1 == df2
ValueError: Can only compare identically-labeled (both index and columns) DataFrame objects
```

Reply:
meaning that the present modin's logic is completely valid for the public versions of the boolean methods and invalid for their private equivalents, ahh...

Reply:
So, it seems like a thing for a different pr: #6029

Reply:
@dchigarev good research, thanks!